src/core/mac/mac_frame.cpp

```diff
@@ -886,25 +893,50 @@ uint8_t Frame::FindHeaderIeIndex(void) const
     return index;
 }
 
-otError Frame::AppendHeaderIe(HeaderIe *aIeList, uint8_t aIeCount)
+#if OPENTHREAD_CONFIG_MAC_HEADER_IE_SUPPORT
+template <typename T> otError Frame::AppendHeaderIeAt(uint8_t &aIndex)
 {
-    otError  error = OT_ERROR_NONE;
-    uint16_t index = FindHeaderIeIndex();
-
-    VerifyOrExit(index != kInvalidIndex, error = OT_ERROR_NOT_FOUND);
+    otError error = OT_ERROR_NONE;
 
-    for (uint8_t i = 0; i < aIeCount; i++)
+    if (aIndex == 0)
     {
-        memcpy(&mPsdu[index], &aIeList[i], sizeof(HeaderIe));
-
-        index += sizeof(HeaderIe) + aIeList[i].GetLength();
-        mLength += sizeof(HeaderIe) + aIeList[i].GetLength();
+        aIndex = FindHeaderIeIndex();
     }
 
+    VerifyOrExit(aIndex != kInvalidIndex, error = OT_ERROR_NOT_FOUND);
+
+    reinterpret_cast<HeaderIe *>(mPsdu + aIndex)->Init(T::kHeaderIeId, sizeof(T));
+    aIndex += sizeof(HeaderIe);
+
+    InitIeContentAt<T>(aIndex);
+
+    mLength += sizeof(HeaderIe) + sizeof(T);
+
```

Comment:
A suggestion here: We can cosnider refactoring to add a common shared method (everything up to `InitIeContentAt<T>()`) as a separate helper (non-template) method.

The main reason is that with the current code we can end with different instantiation of `AppendHeaderIeAt<Type>()` method (each as its own method) repeating similar looking code (the first part up to `InitIeContentAt`), and adding to the overall code size. 

Thoughts?